### DISCUSSION TITLE
Slow formatting of cells

### DISCUSSION THREAD

**Discussion by cvaldezerea** [JanMarvin/openxlsx2#544] [General]
I'm formatting cells with this funcion and it's really slow, am I doing something wrong?

```
openxlsx2::wb_add_numfmt(
      wb = excel,
      sheet = excel_page_name,
      dims = paste0(
        openxlsx2::get_cell_refs(data.frame(starting_row + 1 + 1,starting_col + col_i)),":",
        openxlsx2::get_cell_refs(data.frame(starting_row + 1 + nrow(excel_data),starting_col + col_i))
      ),
      numfmt = "#,##0.0"
    )
```

**Comment by JanMarvin** [comment]
Hi @cvaldezerea , that's hard to tell without knowledge of the actual data and without knowledge of what you consider really slow. Something like "A1:Z10000" on a file without much formatting should work quick.

[Edit:] Just an example of how quick it can be if all cells are initialized
``` r
library(openxlsx2)

mm <- matrix(NA, ncol = 26, nrow = 10000)

beg <- Sys.time()
wb <- wb_workbook()$add_worksheet()
wb$add_data(x = mm, colNames = FALSE, na.strings = NULL)
wb$add_numfmt(dims = "A1:Z10000", numfmt = "#,##0.0")
end <-  Sys.time()
end - beg
#> Time difference of 0.6836898 secs
```

Could you provide some more insight?

**Reply by JanMarvin** [reply]
Okay, I see what you mean. Not sure if this is a regression.  I'll look into it.

**Reply by JanMarvin** [reply]
Okay, here is what I found and I wasn't aware of this, thanks for bringing this to my attention.

Below is a simple task to write a workbook and to set the number format just the way you want to above.

As we can see in the examples below, if all cells are initialized, the numformat application is fast, instead in the second example not all cells are available and we take almost 10 times as long.

If not all cells are available, we have to come up with a quicker way to initialize the cells. Therefore If you have full control over the workbook, you could write an empty matrix just like in the example below and the code should run a lot quicker.

``` r
library(openxlsx2)

mm <- matrix(NA, ncol = 26, nrow = 10000)

# write empty data
beg <- Sys.time()
wb <- wb_workbook()$add_worksheet()
wb$add_data(x = mm, colNames = FALSE, na.strings = NULL)
wb$add_numfmt(dims = "A1:Z100", numfmt = "#,##0.0")
end <-  Sys.time()
end - beg
#> Time difference of 0.4144909 secs

# do not write empty data
beg <- Sys.time()
wb <- wb_workbook()$add_worksheet()
wb$add_numfmt(dims = "A1:Z100", numfmt = "#,##0.0")
end <-  Sys.time()
end - beg
#> Time difference of 4.061839 secs
```

**Reply by JanMarvin** [reply]
This takes a lot of time: 

```R
      # initialize cell
      if (!all(exp_cells %in% got_cells)) {

        init_cells <- NA
        missing_cells <- exp_cells[!exp_cells %in% got_cells]

        for (exp_cell in missing_cells) {
          self$add_data(
            x = init_cells,
            na.strings = NULL,
            colNames = FALSE,
            dims = exp_cell
          )
        }
      }
```

**Reply by cvaldezerea** [reply]
Thank you for the example, I'll try to implement as you do

I was using

```
openxlsx2::wb_add_numfmt(
      wb = excel,
      sheet = excel_page_name,
      dims = paste0(
        openxlsx2::get_cell_refs(data.frame(starting_row + 1 + 1,starting_col + col_i)),":",
        openxlsx2::get_cell_refs(data.frame(starting_row + 1 + nrow(excel_data),starting_col + col_i))
      ),
      numfmt = "#,##0.0"
    )
```
But as you do, i should it like this

```
excel$add_numfmt(dims = paste0(
        openxlsx2::get_cell_refs(data.frame(starting_row + 1 + 1,starting_col + col_i)),":",
        openxlsx2::get_cell_refs(data.frame(starting_row + 1 + nrow(excel_data),starting_col + col_i))
      ), numfmt = "#,##0.0")
```

**Reply by JanMarvin** [reply]
That should not really matter, they are both valid ways. But instead of `get_cell_refs()` I would use something like this (untested draft):

```R
col <- int2col(starting_col + col_i)
beg_row <- starting_row + 1 + 1
end_row <- starting_row + 1 + nrow(excel_data)
dims <- sprintf("%s%s:%s%s, col, beg_row, col, end_row)
```

**Reply by JanMarvin** [reply]
@cvaldezerea was your initial issue solved with the changes in the main branch? I want to create a CRAN release sometime next week.
